### PR TITLE
topo: Move to allow list for invalid names

### PR DESCRIPTION
**Pull request by dbussink**
Today we explicit disallow only specific characters like `/`. We should move here to an allow list approach though, explicitly allowing specific characters.

The reason is that a block list approach is bound to end up in a game of whack-a-mole where we keep having to add characters. Instead an allow list approach of safe characters is a fundamentally better method moving forward that provides security.

The biggest potential downside is that there's people we don't know about that are using names we would disallow with the current restrictions.

From a security perspective though, it's better that we start off here and then allow more known safe things if necessary and try to get feedback from the community for when things are too restricted.

## Related Issue(s)

Followup to https://github.com/vitessio/vitess/pull/12843 & https://github.com/vitessio/vitess/pull/12732 where we added the first validations. This is the more fundamental fix for #12738

I have also marked it for backporting as we've backported the previous changes as well and classified them as a security issue.

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/topo/validator.go
Comment (dbussink) [inline go/vt/topo/validator.go:13]: @derekperkins @arthurschreiber @timvaillancourt @hkdsun I'd love to also explicitly solicit here feedback from outside PlanetScale to see if this suggest validation is problematic for anyone in the community and if we should further loosen it with more known to work characters or character classes or not then.

Comment (timvaillancourt) [reply go/vt/topo/validator.go:13]: @dbussink agreed on the "allowlist" approach

> // They are only allowed to use ASCII letters or digits, - and _.
> // No spaces or special characters are allowed.

Sounds like a reasonable limitation to me 👍

Comment (derekperkins) [reply go/vt/topo/validator.go:13]: Seems reasonable. I'd check with some of the Chinese adopters, as they're more likely to have utf8 characters.

Comment (dbussink) [reply go/vt/topo/validator.go:13]: @derekperkins Do you know their GitHub handles? We can ping them here as well I think :smile:.

Comment (hkdsun) [reply go/vt/topo/validator.go:13]: We have a similar validation on keyspace names on our side. We only use Kubernetes and DNS friendly names 👍

Comment (arthurschreiber) [reply go/vt/topo/validator.go:13]: Sounds fine to me as well! 👍

**Review by deepthi — APPROVED** [review]
Copy/paste error in the release notes needs to be fixed, rest LGTM

**Comment by poddarh** [comment]
We use `.` (dots/periods) pretty extensively in both our keyspace and shard names. Could we add this character to the allow-list as well? Thanks :)

**Comment by dbussink** [comment]
> We use `.` (dots/periods) pretty extensively in both our keyspace and shard names. Could we add this character to the allow-list as well? Thanks :)

@poddarh Hmm, don't you then run into issues often because the `.` is also used as a separate between potential database / keyspace names and the actual tables? Or do you make sure that you quote everything always?

**Comment by poddarh** [comment]
@dbussink I guess we never run into that issue as we don't actually allow customers to query across databases today. We can replace `.` in our shard and keyspace names to another character/set of characters to do this, or just ask the customers to quote things properly :)

**Comment by dbussink** [comment]
> @dbussink I guess we never run into that issue as we don't actually allow customers to query across databases today. We can replace `.` in our shard and keyspace names to another character/set of characters to do this, or just ask the customers to quote things properly :)

It's not just queries across databases, it also affects potentially things like the `vschema` etc. See for example https://github.com/vitessio/vitess/pull/12577 which included fixes for using `-`. 

@poddarh Would either `_` or `-` as a separator work for you? Those are more commonly used by others as well which also means you're not on your own to discover potential issues.

**Comment by poddarh** [comment]
> > @dbussink I guess we never run into that issue as we don't actually allow customers to query across databases today. We can replace `.` in our shard and keyspace names to another character/set of characters to do this, or just ask the customers to quote things properly :)
> 
> It's not just queries across databases, it also affects potentially things like the `vschema` etc. See for example #12577 which included fixes for using `-`.
> 
> @poddarh Would either `_` or `-` as a separator work for you? Those are more commonly used by others as well which also means you're not on your own to discover potential issues.

We only stared using this recently and is only available as a prototype internally as we evaluate whether it meets our needs or not. I will look into replacing `.` with some other sequence/characters such that they don't conflict with other shard/keyspace names. Thanks for sharing these options and suggestions :)

File: changelog/17.0/17.0.0/summary.md
Comment (ajm188) [reply changelog/17.0/17.0.0/summary.md:82]: This is still incorrect

**Review by ajm188 — APPROVED** [review]
LGTM pending application of @deepthi's suggested change

File: go/vt/topo/shard.go
Comment (ajm188) [inline go/vt/topo/shard.go:300]: kinda nitpicky, but maybe we should validate the shard name first as well, so we can skip locking entirely if we're just going to fail the call?

Comment (dbussink) [reply go/vt/topo/shard.go:306]: Ah yeah, let's do that. Fixed!

**Pull request merged**
